### PR TITLE
Core/Movement: Use time synchronisation to better interpret client timestamps

### PR DESCRIPTION
**Pull request by chaodhib**
Hello,

This is a port of [the time synchronization mechanism implemented in TrinityCore](https://github.com/TrinityCore/TrinityCore/pull/18189). The biggest hurdle for this port was to replace the usage of Boost. 2 utility files (MathUtil.h and CircularBuffer.h) could be removed in the future, once (if) Boost is added as a dependency to the project.

**Changes proposed:** _Copied from the original PR_

- Implement [Simple Network Time Protocol](https://tools.ietf.org/html/rfc4330) (SNTP) using the `SMSG_TIME_SYNC_REQUEST` /  `CMSG_TIME_SYNC_RESPONSE` exchange. No need to read however the RFC. The mechanism is quite simple and it is explained in the code. This packet exchange initiated by the server provides to the server the following information: the difference between its own clock and the clock of the client. With that clock delta, the server can precisely interpret timestamps sent by the client.
- On top of SNTP, I implemented the approach described [HERE](https://web.archive.org/web/20180430214420/http://www.mine-control.com/zack/timesync/timesync.html). That approach is necessary to use on top of SNTP because plain SNTP is designed to work on UDP, not TCP. The fact that TCP can resend dropped packets without any warning requires extra care.
- Using the server newly found information, which is the clock delta, we can now interpret the only timestamps sent by the client to the server: The movement timestamps. Now, when movement events occur on the client, we can precisely know their time on server clock (such as the player has started to move forward, for example). And send these movement timestamps on server clock to the observing nearby players.

**Issues addressed:** It should improve the visual perception of the other players on screen. 

## SOURCE:
- https://github.com/TrinityCore/TrinityCore/pull/18189

## Tests Performed:

- Tested in game locally.

## How to Test the Changes:
- By using (at least) 2 clients and observing other players move. The movement of other players should appear fluid with as few visual artifacts (such as abnormal teleports and going back of the character) as possible. Ideally, this PR should decrease the number of visual artifacts compared to the current master.

## Known Issues and TODO List:

- [ ]
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
Very nice

File: src/common/Utilities/CircularBuffer.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
+/*
+ * This file was based on
+ * https://embeddedartistry.com/blog/2017/05/17/creating-a-circular-buffer-in-c-and-c/
+ * https://github.com/embeddedartistry/embedded-resources/blob/master/examples/cpp/circular_buffer.cpp
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
+ */
```
Comment (Winfidonarleyan) [inline src/common/Utilities/CircularBuffer.h:8]:
```cpp
/*
 * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
 *
 * This file was based on
 * https://embeddedartistry.com/blog/2017/05/17/creating-a-circular-buffer-in-c-and-c/
 * https://github.com/embeddedartistry/embedded-resources/blob/master/examples/cpp/circular_buffer.cpp
 */
```

File: src/common/Utilities/MathUtil.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
+ */
```
Comment (Winfidonarleyan) [inline src/common/Utilities/MathUtil.h:5]:
```cpp
/*
 * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
 */
```

File: src/server/game/Server/WorldSocket.cpp [hunk 3]
```diff
@@ -727,6 +710,26 @@ int WorldSocket::ProcessIncoming(WorldPacket* new_pct)
         return -1;
     }
 
+    ACE_GUARD_RETURN(LockType, Guard, m_SessionLock, -1);
+
+    if (m_Session != nullptr)
+    {
+        // Our Idle timer will reset on any non PING or TIME_SYNC opcodes.
+        // Catches people idling on the login screen and any lingering ingame connections.
+        if(opcode != CMSG_PING && opcode != CMSG_TIME_SYNC_RESP)
```
Comment (Winfidonarleyan) [inline src/server/game/Server/WorldSocket.cpp:719]: after `if` need add whitespace

**Comment by Winfidonarleyan** [comment]
@chaodhib Big thanks for PR.

About license
- For new files need add
```cpp
/*
 * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
 */
```

**Comment by Footman** [comment]
tested on live. everything fine

**Comment by chaodhib** [comment]
@Footman thank you for testing.
@Winfidonarleyan thank you for making the necessary fixes to the headers and codestyle. I have added the missing section (how to test).

**Review by Yehonal — APPROVED** [review]

**Comment by Yehonal** [comment]
Keep up the Good Work ™️

File: src/server/game/Server/WorldSocket.cpp [hunk 4]
```diff
@@ -724,6 +706,28 @@ int WorldSocket::ProcessIncoming(WorldPacket* new_pct)
...
+        m_Session->QueuePacket(new_pct);
+        return 0;
+    }
+    else
+    {
+        LOG_ERROR("server", "WorldSocket::ProcessIncoming: Client not authed opcode = %u", uint32(opcode));
+        return -1;
+    }
+
     ACE_NOTREACHED (return 0);
```
Comment (Helias) [inline src/server/game/Server/WorldSocket.cpp:731]: this will never happen

File: src/server/game/Server/WorldSocket.cpp [hunk 5]
```diff
@@ -724,6 +706,28 @@ int WorldSocket::ProcessIncoming(WorldPacket* new_pct)
...
+        // OK, give the packet to WorldSession
+        aptr.release();
+        m_Session->QueuePacket(new_pct);
+        return 0;
+    }
+    else
+    {
+        LOG_ERROR("server", "WorldSocket::ProcessIncoming: Client not authed opcode = %u", uint32(opcode));
+        return -1;
+    }
```
Comment (Helias) [inline src/server/game/Server/WorldSocket.cpp:729]:
```suggestion
    LOG_ERROR("server", "WorldSocket::ProcessIncoming: Client not authed opcode = %u", uint32(opcode));
    return -1;
```

**Review by Helias — CHANGES_REQUESTED** [review]

**Comment by FrancescoBorzi** [comment]
@chaodhib before merging this, can you please check them comments from @Helias and see if they make sense to you?

**Comment by chaodhib** [comment]
Sure, will do

File: src/server/game/Handlers/MovementHandler.cpp [hunk 6]
```diff
@@ -796,3 +800,76 @@ void WorldSession::HandleMoveTimeSkippedOpcode(WorldPacket& recvData)
...
+        return;
+
+    uint32 serverTimeAtSent = _pendingTimeSyncRequests.at(counter);
+    _pendingTimeSyncRequests.erase(counter);
+
+    // time it took for the request to travel to the client, for the client to process it and reply and for response to travel back to the server.
+    // we are going to make 2 assumptions:
+    // 1) we assume that the request processing time equals 0.
+    // 2) we assume that the packet took as much time to travel from server to client than it took to travel from client to server.
+    uint32 roundTripDuration = getMSTimeDiff(serverTimeAtSent, recvData.GetReceivedTime());
```
Comment (pklloveyou) [inline src/server/game/Handlers/MovementHandler.cpp:823]: Errors are reported here

File: src/server/game/Handlers/MovementHandler.cpp [hunk 7]
```diff
@@ -796,3 +800,76 @@ void WorldSession::HandleMoveTimeSkippedOpcode(WorldPacket& recvData)
...
+    ComputeNewClockDelta();
+}
+
+void WorldSession::ComputeNewClockDelta()
+{
+    // implementation of the technique described here: https://web.archive.org/web/20180430214420/http://www.mine-control.com/zack/timesync/timesync.html
+    // to reduce the skew induced by dropped TCP packets that get resent.
+
+    vector<uint32> latencies;
+    vector<int64> clockDeltasAfterFiltering;
```
Comment (pklloveyou) [inline src/server/game/Handlers/MovementHandler.cpp:847]: std::vector<uint32> latencies;
    std::vector<int64> clockDeltasAfterFiltering;

**Comment by pklloveyou** [comment]
image
It's really much better

File: src/common/Utilities/CircularBuffer.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    // the implementation of this function is simplified by the fact that head_ will never be lower than tail_
+    std::vector<T> content() {
+        std::lock_guard<std::mutex> lock(mutex_);
+
+        if(full_)
+            return std::vector<T>(buf_.get(), buf_.get() + max_size_);
+        else
+            return std::vector<T>(buf_.get(), buf_.get() + head_ - tail_);
```
[the remark this answers is not in the document]
Comment (chaodhib) [reply src/common/Utilities/CircularBuffer.h]: I changed to something even shorter, thank you for the review

File: src/server/game/Server/WorldSocket.cpp (hunk 4 above)
Comment (chaodhib) [reply src/server/game/Server/WorldSocket.cpp]: Removed

File: src/server/game/Server/WorldSocket.cpp (hunk 5 above)
Comment (chaodhib) [reply src/server/game/Server/WorldSocket.cpp]: Done

File: src/server/game/Handlers/MovementHandler.cpp (hunk 7 above)
Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp]: Done

**Comment by chaodhib** [comment]
Done. Let me know if there is something else that could be improved.

@pklloveyou thanks for the feedback! Appreciated!

**Comment by pklloveyou** [comment]
Tested

**Comment by FrancescoBorzi** [comment]
good job @chaodhib ! I will merge this as soon as the build passes.

PS: would you like to join our [Discord chat](https://discordapp.com/invite/gkt4y2x)?

**Comment by chaodhib** [comment]
@FrancescoBorzi Thank you! Sure, I'm in the Discord already. My handle is #7251.

**Comment by Winfidonarleyan** [comment]
Thanks for pr

**Comment by FrancescoBorzi** [comment]
> @FrancescoBorzi Thank you! Sure, I'm in the Discord already. My handle is #7251.

@chaodhib  what's your username? I cannot find you

**Comment by chaodhib** [comment]
Chaouki

**Pull request merged**